### PR TITLE
Tensor Metrics 2: Expose MCore tensors through observation hooks

### PR DESCRIPTION
**Pull request by philipcmonk**
Adds dormant, scoped MCore hooks that expose selected short-lived tensors to an external observer without depending on the training tensor-metric implementation.  These will be used in a later training-focused PR.

This is part 2 of a four-PR stack.  However, this and PR 1 are completely independent and can be reviewed/merged in either order (just rebase this onto main if this is to be merged first).

1. #6647: TensorMetric core framework
2. **This PR:** MCore tensor observation sources
3. #6649: metric definitions
4. #6650: training-loop integration

Some useful training diagnostics require tensors that do not survive until the optimizer step, including residual streams and router or output logits.

MCore should expose these values with minimal instrumentation while leaving metric selection, distributed reduction, scheduling, and logging outside `megatron/core`.

## Design

`megatron.core.tensor_observation` provides a scoped callback mechanism based on a context variable. Outside an active observation scope, every hook is a cheap no-op.

The callback receives:

- the owning module;
- a name local to that owner;
- a stable "source kind";
- the tensor;
- its optional tensor-parallel shard dimension.

The source hook does not import or understand TensorMetric classes.

## Observation sites

This PR exposes:

- residual accumulators;
- net per-layer residual contributions;
- GPT and HybridModel output logits;
- MTP logits;
- raw router logits before forced load-balancing modifications;
- normalized router decision scores;
- compact per-sequence router diagnostics.

We will later want to expose more observation sites, but this is the list of sites that we have so far used for metrics.  More can be added when needed.

## Canonical parameter names

This PR also adds `CanonicalParameterNameMap`, which creates stable logical names across:

- pipeline-local layer numbering;
- MTP layers;
- grouped and sequential experts;
- expert-parallel ownership;
- tied parameters and model chunks.

The logic included here is replicated ad hoc in several places in the codebase, so it seems worth centralizing it here.  However, this could be moved into training/tensor_metrics if needed.

### CODE REVIEW

**Comment by philipcmonk** [comment]
I'm not sure why that Claude run didn't produce a review: https://github.com/NVIDIA/Megatron-LM/actions/runs/32337268716/job/96329132796

```
  {
    "type": "result",
    "subtype": "success",
    "is_error": false,
    "duration_ms": 8648182,
    "num_turns": 137,
    "total_cost_usd": 22.5640935,
    "permission_denials_count": 20
  }
```

I'll try running it again.

File: megatron/core/extensions/transformer_engine.py [hunk 1]
```diff
@@ -3374,9 +3374,22 @@ def te_checkpoint(
...
+    def forward_func_without_recomputed_observations(*forward_args, **forward_kwargs):
+        nonlocal initial_forward
+        if initial_forward:
+            initial_forward = False
+            return forward_func(*forward_args, **forward_kwargs)
+
+        from megatron.core.tensor_observation import suspend_tensor_observations
+
+        with suspend_tensor_observations():
+            return forward_func(*forward_args, **forward_kwargs)
```
[the remark this answers is not in the document]
Comment (philipcmonk) [reply megatron/core/extensions/transformer_engine.py]: The suspension is safe whether the autograd worker inherits Python context or not:
- If it inherits the active observer, it also sees the suspension set in that same context.
- If it starts with an empty context, it cannot see the observer in the first place.

File: megatron/core/parameter_names.py [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+
+from megatron.core.utils import unwrap_model
+
+_GROUPED_EXPERT_PATTERN = re.compile(
+    r"^((?:.*\.)?mlp\.experts\.linear_fc\d\.(?:weight|bias))(\d+)(.*)$"
+)
+_SEQUENTIAL_EXPERT_PATTERN = re.compile(r"^((?:.*\.)?mlp\.experts\.local_experts\.)(\d+)(\..*)$")
+
+
+class CanonicalParameterNameMap(Mapping[torch.nn.Parameter, str]):
```
[the remark this answers is not in the document]
Comment (philipcmonk) [reply megatron/core/parameter_names.py:23]: Those later PRs are part of this stack, this was explicitly split up to keep separate the MCore changes from the broader integration into training/

File: megatron/core/tensor_observation.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+    if is_observing_tensor("residual_accumulator"):
+        _notify_observer(
+            layer, "residual_accumulator", "residual_accumulator", accumulator, tp_shard_dim
+        )
+    if is_observing_tensor("residual_contribution"):
+        with torch.no_grad():
+            contribution = output.detach() - accumulator.detach()
+        _notify_observer(
+            layer, "residual_contribution", "residual_contribution", contribution, tp_shard_dim
+        )
```
[the remark this answers is not in the document]
Comment (philipcmonk) [reply megatron/core/tensor_observation.py:121]: This is worth changing the docstring, but the assertion that we're not using full_iteration cudagraphs is in #6650 and probably shouldn't be repeated.

File: megatron/core/models/gpt/gpt_model.py [hunk 4]
```diff
@@ -771,6 +772,18 @@ def _postprocess(
...
+        gather_output = (
+            getattr(self.output_layer, "gather_output", False)
+            if runtime_gather_output is None
+            else runtime_gather_output
+        )
+        observe_tensor(
+            self.output_layer,
+            "output_logits",
+            "output_logits",
+            logits,
```
[the remark this answers is not in the document]
Comment (philipcmonk) [reply megatron/core/models/gpt/gpt_model.py:784]: Agree that the `getattr` is wrong, will fix.  But I don't know that the gather_output behavior will always be the same between those three call-sites, it doesn't seem like enough duplication to be worth the helper.

File: megatron/core/tensor_parallel/random.py [hunk 5]
```diff
@@ -655,7 +655,9 @@ def backward(ctx, *args):
 
             # Compute the forward pass.
             detached_inputs = detach_variable(inputs)
-            with torch.enable_grad():
+            from megatron.core.tensor_observation import suspend_tensor_observations
+
+            with torch.enable_grad(), suspend_tensor_observations():
```
[the remark this answers is not in the document]
Comment (philipcmonk) [reply megatron/core/tensor_parallel/random.py:659]: The suspension is safe whether the autograd worker inherits Python context or not:

- If it inherits the active observer, it also sees the suspension set in that same context.
- If it starts with an empty context, it cannot see the observer in the first place.

File: megatron/core/extensions/transformer_engine.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (philipcmonk) [reply megatron/core/extensions/transformer_engine.py:3387]: It's unclear under what circumstances we would ever call it three times.  Even so, this doesn't assume that it's called exactly twice, it observes the first call and no subsequent calls.  Checking `torch.is_grad_enabled()` is less explicit and more tied to how things happen to be implemented right now.

Re: thread-safety, the first call is synchronous, so by the time any subsequent calls start, `initial_forward` is already false.

File: megatron/core/transformer/moe/router.py [hunk 6]
```diff
@@ -798,36 +816,53 @@ def routing(self, logits: torch.Tensor, padding_mask: Optional[torch.Tensor] = N
...
+            observe_router_diagnostics or (torch.is_grad_enabled() and aux_loss_enabled)
+        )
+        if compute_aux_inputs:
             routing_map_for_aux_loss, scores_for_aux_loss = compute_routing_scores_for_aux_loss(
                 logits,
                 self.topk,
                 self.score_function,
                 fused=self.config.moe_router_fusion,
                 padding_mask=padding_mask,
             )
```
[the remark this answers is not in the document]
Comment (philipcmonk) [reply megatron/core/transformer/moe/router.py]: Agree, fixed in f4b97a4

File: megatron/core/parameter_names.py [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+from __future__ import annotations
+
+import re
+from collections.abc import Iterable, Iterator, Mapping
+
+import torch
+
+from megatron.core.utils import unwrap_model
+
+_GROUPED_EXPERT_PATTERN = re.compile(
```
[the remark this answers is not in the document]
Comment (philipcmonk) [reply megatron/core/parameter_names.py:14]: Agree, fixed in f4b97a4

File: megatron/core/transformer/moe/router.py [hunk 8]
```diff
@@ -855,6 +890,24 @@ def forward(self, input: torch.Tensor, padding_mask: Optional[torch.Tensor] = No
...
+                    scores = torch.sigmoid(logits.float())
+                    scores = scores / (scores.sum(dim=-1, keepdim=True) + 1e-20)
+                elif self.score_function == "sqrtsoftplus":
+                    scores = torch.nn.functional.softplus(logits.float()).sqrt()
+                    scores = scores / (scores.sum(dim=-1, keepdim=True) + 1e-20)
+                else:
+                    raise ValueError(f"Invalid score_function: {self.score_function}")
+            observe_tensor(
+                self, "router_scores", "router_scores", scores, tp_shard_dim=tp_shard_dim
+            )
```
[the remark this answers is not in the document]
Comment (philipcmonk) [reply megatron/core/transformer/moe/router.py:924]: Agree, fixed in f4b97a4

File: megatron/core/transformer/moe/router_diagnostics.py [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+    has_boundary = aux_map.any(dim=-1) & (~aux_map).any(dim=-1) & valid_mask
+    selected_floor = scores.masked_fill(~aux_map, float("inf")).amin(dim=-1)
+    unselected_ceiling = scores.masked_fill(aux_map, float("-inf")).amax(dim=-1)
+    selected_floor = torch.where(has_boundary, selected_floor, torch.ones_like(selected_floor))
+    unselected_ceiling = torch.where(has_boundary, unselected_ceiling, selected_floor)
+    boundary_margin = (selected_floor - unselected_ceiling).clamp_min(0)
+    boundary_relative_margin = boundary_margin / selected_floor.clamp_min(
+        torch.finfo(scores.dtype).eps
+    )
+    boundary_relative_margin = boundary_relative_margin.sum(dim=0) / has_boundary.sum(
```
[the remark this answers is not in the document]
Comment (philipcmonk) [reply megatron/core/transformer/moe/router_diagnostics.py:109]: Agree, fixed in f4b97a4

File: megatron/core/tensor_parallel/random.py [hunk 10]
```diff
@@ -655,7 +655,9 @@ def backward(ctx, *args):
 
             # Compute the forward pass.
             detached_inputs = detach_variable(inputs)
```
[the remark this answers is not in the document]
Comment (philipcmonk) [reply megatron/core/tensor_parallel/random.py:658]: Agree, fixed in f4b97a4

**Comment by philipcmonk** [comment]
Claude did the [same thing](https://github.com/NVIDIA/Megatron-LM/actions/runs/32426158607/job/96608421760) as before, ran for 2.5 hours then ended "successfully" without posting a review.

**Comment by Phlip79** [comment]
> Claude did the [same thing](https://github.com/NVIDIA/Megatron-LM/actions/runs/32426158607/job/96608421760) as before, ran for 2.5 hours then ended "successfully" without posting a review.

Based off my findings, the GitHub token has a 1 hour expiration from job start time. Anyways, 2.5 hour runtime is way too high.

**Comment by philipcmonk** [comment]
Added that test in [51c1d10](https://github.com/NVIDIA/Megatron-LM/pull/6648/commits/51c1d100a8284f66cde40cdca3204b6b9d186985).  Re (2) we don't support pipeline parallelism with this work yet.

**Review by yaox12 — APPROVED** [review]
LGTM for the MoE part.

File: megatron/core/tensor_observation.py [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+    Python observation notifications. The contribution is computed without autograd only when a
+    contribution metric is active.
+
+    Args:
+        layer: Layer owning the residual sites.
+        accumulator: Residual stream entering the layer.
+        output: Residual stream leaving the layer.
+    """
+    config = getattr(layer, "config", None)
+    tp_shard_dim = 0 if getattr(config, "sequence_parallel", False) else None
```
Comment (janEbert) [inline megatron/core/tensor_observation.py:105]: 1. This is missing context-parallelism handling, isn't it? Should we error similar to `router.py` if CP>1?
2. Tests exercising `observe_layer_residuals` with TP>1 and/or CP>1 on a `TransformerBlock`/`HybridStack` would be nice to catch issues going forward.
3. This assumes a `tp_shard_dim` if `sequence_parallel=True`, even if TP=1.

Comment (philipcmonk) [reply megatron/core/tensor_observation.py:105]: If `sequence_parallel=True`, then TP is enforced here to be >1: https://github.com/NVIDIA/Megatron-LM/blob/a97b0e20315e03d9a59fc1001311fb2c832febef/megatron/core/model_parallel_config.py#L502-L504

For CP, technically the fallback to `None` is acceptable (though should be tested), because `None` means "it's sharded along this process group, but we're not specifying exactly how".  However, I think we can and should be more precise -- CP is sharded along dim 0.  We're also sharding along DP and GTP as `None`, and that really should be dim 1 as well, since these have shape `[sequence, batch, dim]`.

I did a pass over this to be more precise about sharded dimensions when we have enough information (which is most of the time).  Also added a test for the CP>1 case.  See [a806c2f](https://github.com/NVIDIA/Megatron-LM/pull/6648/commits/a806c2f8b846e19e2090e6119f18bb81523cbe15)

Comment (janEbert) [reply megatron/core/tensor_observation.py:105]: Thank you!

**Review by zhongbozhu — APPROVED** [review]

**Review by janEbert — APPROVED** [review]

**Review by guihong-nv — APPROVED** [review]

File: megatron/core/transformer/moe/router.py [hunk 12]
```diff
@@ -756,6 +760,22 @@ def routing(self, logits: torch.Tensor, padding_mask: Optional[torch.Tensor] = N
                 with shape [num_tokens, num_experts].
         """
         seq_length, bsz = logits.shape[:2]
+        observe_router_diagnostics = is_observing_tensor("router_diagnostics")
```
Comment (yashaswikarnati) [inline megatron/core/transformer/moe/router.py:770]: when we do THD / no inter document masking, i wonder if the metrics we capture here needs to take into account per-sequence

File: megatron/core/transformer/moe/router.py [hunk 13]
```diff
@@ -798,36 +818,60 @@ def routing(self, logits: torch.Tensor, padding_mask: Optional[torch.Tensor] = N
...
+                    self.score_function,
+                    fused=self.config.moe_router_fusion,
+                    padding_mask=padding_mask,
+                )
+            if observe_router_diagnostics:
+                diagnostics = build_router_diagnostics(
+                    scores_for_aux_loss,
+                    routing_map_for_aux_loss,
+                    routing_map,
+                    self.expert_bias,
```
Comment (yashaswikarnati) [inline megatron/core/transformer/moe/router.py]: codex flagged passing self.expert bias here might be incorrec for quantile balancing

File: megatron/core/transformer/moe/router.py [hunk 14]
```diff
@@ -855,6 +899,29 @@ def forward(self, input: torch.Tensor, padding_mask: Optional[torch.Tensor] = No
...
+            logits,
+            tp_shard_dim=tp_shard_dim,
+            sequence_dim=0,
+            batch_dim=1,
+        )
+        # Materialize the full decision distribution only when a due metric requests it.
+        if is_observing_tensor("router_scores"):
+            with torch.no_grad():
+                scores = compute_normalized_router_scores(logits, self.score_function)
+            observe_tensor(
```
Comment (yashaswikarnati) [inline megatron/core/transformer/moe/router.py]: should we also pass in padding mask and only account for valid tokens

**Review by Phlip79 — APPROVED** [review]

**Review by guihong-nv — APPROVED** [review]

File: megatron/core/transformer/moe/router.py (hunk 12 above)
Comment (philipcmonk) [reply megatron/core/transformer/moe/router.py:770]: Yes, right now all of these are per-sample not per-document.  At some point it would be nice to support THD, but I think it's fine to defer for now?

File: megatron/core/transformer/moe/router.py (hunk 13 above)
Comment (philipcmonk) [reply megatron/core/transformer/moe/router.py]: Good catch, I wasn't looking at quantile balancing at the time.  Fixed in b7a87c9df6a1931800213e4be64b66862742df11

File: megatron/core/transformer/moe/router.py (hunk 14 above)
Comment (philipcmonk) [reply megatron/core/transformer/moe/router.py]: Yes, fixed in b7a87c9df6a1931800213e4be64b66862742df11

File: megatron/core/transformer/moe/router.py (hunk 12 above)
Comment (yashaswikarnati) [reply megatron/core/transformer/moe/router.py:770]: sgtm

**Review by yashaswikarnati — APPROVED** [review]

**Review by jaredcasper — APPROVED** [review]

**Comment by chtruong814** [comment]
Ran CI as part of an aggregated branch that included multiple PRs in the queue.
https://github.com/NVIDIA/Megatron-LM/pull/7459

CI passed
https://github.com/NVIDIA/Megatron-LM/actions/runs/35291473099?pr=7459

Will fast merge this
